{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyOym+YRu92kaR63sTZfjhAr"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","source":["# Source : https://www.reneshbedre.com/blog/dbscan-python.html\n","# DBSCAN Video : https://www.youtube.com/watch?v=RDZUdRSDOok\n","\n","# For clustering using DBSCAN, we are using a single-cell gene expression dataset \n","# of Arabidopsis thaliana root cells processed by a 10x genomics Cell Ranger pipeline. \n","# The dataset is preprocessed t-SNE dimensionality reduction technique. \n","# Now, we will use the t-SNE embedding vectors to identify the clusters using DBSCAN."],"metadata":{"id":"chxWFzsVhPEP"},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"id":"msgE1oNSWH5y"},"outputs":[],"source":["import pandas as pd\n","import numpy as np\n","import math\n","\n","from sklearn.neighbors import NearestNeighbors\n","\n","from collections import Counter\n","\n","from sklearn.cluster import KMeans\n","from sklearn.cluster import AgglomerativeClustering\n","from sklearn.cluster import DBSCAN\n","\n","\n","import scipy.cluster.hierarchy as shc\n","\n","import seaborn as sns\n","import matplotlib.pyplot as plt\n","import matplotlib\n","\n","import plotly.express as px\n","import plotly.graph_objs as go"]},{"cell_type":"code","source":["df = pd.read_csv(\"https://reneshbedre.github.io/assets/posts/tsne/tsne_scores.csv\")\n","# df = pd.read_csv(\"SelfShiksha_ULB_FC89_DBSCAN.csv\")"],"metadata":{"id":"zVdtlyI6VWro"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["df"],"metadata":{"id":"IA2UjEUVVWue"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# To determine the optimal eps parameter of DBSCAN algorithm, \n","# it is common practice to compute the k-nearest neighbor (kNN) distances \n","# (average distance of every data point to its k-nearest neighbors) \n","# of an input dataset using the k-nearest neighbor method.\n","# NearestNeighbors function requires n_neighbors (number of neighbors) parameter, which can be same as the minPts value.\n","\n","# n_neighbors = 5 as kneighbors function returns distance of point to itself (i.e. first column will be zeros) \n","nbrs = NearestNeighbors(n_neighbors=5).fit(df)\n","# Find the k-neighbors of a point\n","neigh_dist, neigh_ind = nbrs.kneighbors(df)\n","# sort the neighbor distances (lengths to points) in ascending order\n","# axis = 0 represents sort along first axis i.e. sort along row\n","sort_neigh_dist = np.sort(neigh_dist, axis=0)"],"metadata":{"id":"oWsV0qP_VWw2"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# In the k-NN distance plot, you should look for the “knee” of the curve to find the optimal value of eps. \n","# In the below plot, the knee occurs at approximately 2.5 i.e. \n","# the points below 2.5 belong to a cluster and points above 2.5 \n","# are noise or outliers (noise points will have higher kNN distance).\n","\n","k_dist = sort_neigh_dist[:, 4]\n","plt.plot(k_dist)\n","plt.axhline(y=2.5, linewidth=1, linestyle='dashed', color='k')\n","plt.ylabel(\"k-NN distance\")\n","plt.xlabel(\"Sorted observations (4th NN)\")\n","plt.show()"],"metadata":{"id":"97RF5MQ7VWzt"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["clusters = DBSCAN(eps=2.5, min_samples=4).fit(df)\n","# get cluster labels\n","clusters.labels_"],"metadata":{"id":"KLcQahGAVW2U"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# check unique clusters\n","set(clusters.labels_)\n","# -1 value represents noisy points could not assigned to any cluster"],"metadata":{"id":"KUFZCXQyWDfj"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["Counter(clusters.labels_)"],"metadata":{"id":"1vQTT6gXWFDw"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["p = sns.scatterplot(data=df, x=\"t-SNE-1\", y=\"t-SNE-2\", hue=clusters.labels_, legend=\"full\", palette=\"deep\")\n","sns.move_legend(p, \"upper right\", bbox_to_anchor=(1.17, 1.2), title='Clusters')\n","plt.show()"],"metadata":{"id":"fTWDRIOdWKJH"},"execution_count":null,"outputs":[]}]}